db/python/tables/output_file.py

```diff
@@ -0,0 +1,266 @@
+        parent_id: int | None = None,
+        client: Client | None = None,
+    ) -> int | None:
+        """
+        Create a new file, and add it to database
+        """
+        file_obj = AnyPath(path, client=GSClient(storage_client=client))
+
+        if not file_obj or not client:
+            raise ValueError('Invalid file path or client')
+
+        file_info = await OutputFileInternal.get_file_info(
+            file_obj=file_obj, client=client
+        )
+
+        if not file_info or not file_info.get('valid'):
+            return None
+
+        kv_pairs = [
+            ('path', path),
+            ('basename', file_info['basename']),
+            ('dirname', file_info['dirname']),
+            ('nameroot', file_info['nameroot']),
+            ('nameext', file_info['nameext']),
+            ('file_checksum', file_info['checksum']),
+            ('size', file_info['size']),
+            ('valid', file_info['valid']),
+            ('parent_id', parent_id),
+        ]
+
+        kv_pairs = [(k, v) for k, v in kv_pairs if v is not None]
+        keys = [k for k, _ in kv_pairs]
+        cs_keys = ', '.join(keys)
+        cs_id_keys = ', '.join(f':{k}' for k in keys)
+        non_pk_keys = [k for k in keys if k != 'path']
+        update_clause = ', '.join(
+            [f'{k} = VALUES({k})' for k in non_pk_keys]
+        )  # ON DUPLICATE KEY UPDATE {update_clause}
+
+        async with self.connection.transaction():
+            _query = dedent(
+                f"""INSERT INTO output_file ({cs_keys}) VALUES ({cs_id_keys}) ON DUPLICATE KEY UPDATE {update_clause} RETURNING id"""
+            )
+            id_of_new_file = await self.connection.fetch_val(
+                _query,
+                dict(kv_pairs),
+            )
+
+        return id_of_new_file
+
+    async def add_output_file_to_analysis(
+        self,
+        analysis_id: int,
+        file_id: int | None,
+        json_structure: str | None = None,
+        output: str | None = None,
+    ):
+        """Add file to an analysis (through the join table)"""
+        _query = dedent(
+            """
+            INSERT INTO analysis_outputs
+                (analysis_id, file_id, json_structure, output)
+            VALUES (:analysis_id, :file_id, :json_structure, :output)
+        """
+        )
+        await self.connection.execute(
+            _query,
+            {
+                'analysis_id': analysis_id,
+                'file_id': file_id,
+                'json_structure': json_structure,
+                'output': output,
+            },
+        )
+
+    async def create_or_update_analysis_output_files_from_output(
+        self,
+        analysis_id: int,
+        json_dict: RecursiveDict | str,
+    ) -> None:
+        """
+        Create analysis files from JSON
+        """
+        files = await self.find_files_from_dict(json_dict=json_dict)
+        file_ids: list[int] = []
+
+        if os.environ.get('SM_ENVIRONMENT', 'local').lower() in (
```

Comment:
There seems to be a weird behaviour with this in tests because despite having `os.environ['SM_ENVIRONMENT'] = 'test'` in `testbase.py`, the value returned from `api.settings` is `production` and this causes my tests to fail. I've left this as is but worth investigating later on